recirq/qml_lfe/learn_states_q.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2021 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Draw I + P pauli product data using two copy strategy.
+style circuits.
+
+learn_states_q.py --n=6 --n_paulis=20 --batch_size=500 --n_shots=1000
+
+Will create paulistring circuits all on 12 qubits (two systems of size 6).
+Once the circuits are generated, batch_size sweeps and a single circuit at
+a timewill be sent will be sent for simulation/execution, drawing n_shots
+total samples from each one using `run_sweep` in Cirq. By default the
+bitstring data will be saved in the data folder. One can also set
+`use_engine` to True in order to run this against a processor on
+quantum engine.
+"""
+from typing import Dict, List
+
+import os
+import cirq
+import numpy as np
+import sympy
+from . import circuit_blocks
+from . import run_config
+
+from absl import app
+from absl import logging
+
+
+def _create_randomized_sweeps(
+    hidden_p: str, symbols: Dict[str, int], num_reps: int
+) -> List[Dict[str, int]]:
+    """Generate sweeps to help prepare \rho = 2^(-n) (I + \alpha P) states.
+
+    See section A.2.a (https://arxiv.org/pdf/2112.00778.pdf) more details.
+
+    Args:
+        hidden_p: Pauli operator in (I + \alpha P)
+        symbols: Symbols to generate values for to prepare \rho.
+        num_reps: Number of unique parameter configurations (sweeps)
+            to generate for the given `symbols`.
+
+    Returns:
+        List of sweeps, that when placed in a circuit will realize
+        \rho.
+    """
+
+    last_i = 0
+    for i, pauli in enumerate(hidden_p):
+        if pauli != "I":
+            last_i = i
+
+    sign_p = -1 if np.random.random() < 0.5 else 1
+    all_sweeps = []
+    for _ in range(num_reps):
+        current_sweep = dict()
+        for twocopy in [0, 1]:
+            parity = sign_p * (1 if np.random.random() <= 0.95 else -1)
+            for i, pauli in enumerate(hidden_p):
+                current_symbol = symbols[2 * i + twocopy]
+                current_sweep[current_symbol] = 0
+                if pauli != "I":
+                    if last_i == i:
+                        current_sweep[current_symbol] = 1 if parity == -1 else 0
```

Comment:
this code is _very_ highly nested. any way to factor out? or use `np.product` instead of nested for loops?

Reply:
Did my best to un-nest, still not great, but better than before.